tests/unit_tests/test_matchmaker_algorithm_team_matchmaker.py

```diff
@@ -0,0 +1,321 @@
+    c1 = CombinedSearch(*[s.pop(), s.pop(), s.pop()])
+    s.append(c1)
+
+    team_a = CombinedSearch(*[c1, s[3]])
+    team_b = CombinedSearch(*[s[0], s[1], s[2], s[4]])
+    teams = TeamMatchMaker().make_teams(s, 4)
+
+    assert set(teams[0].players) == set(team_a.players)
+    assert set(teams[1].players) == set(team_b.players)
+
+
+def test_ignore_impossible_team_splits(player_factory):
+    s = make_searches([1415, 125, 1294, 584, 1303, 1526, 1237, 1218], player_factory)
+    c1 = CombinedSearch(*[s.pop(), s.pop(), s.pop()])
+    c2 = CombinedSearch(*[s.pop(), s.pop(), s.pop()])
+    c3 = CombinedSearch(*[s.pop(), s.pop()])
+    s.append(c1)
+    s.append(c2)
+    s.append(c3)
+
+    matches, unmatched = TeamMatchMaker().find(s, 4)
+
+    assert len(matches) == 0
+    assert set(unmatched) == set(s)
+
+
+@given(team_a=st_searches(4), team_b=st_searches(4))
+def test_game_quality(team_a, team_b):
+    game = TeamMatchMaker().assign_game_quality((team_a, team_b))
+
+    assert 0.0 <= game.quality <= 1.0
+
+
+@given(player=st_players())
+def test_maximum_game_quality_for_even_teams(player):
+    search = Search([player])
+    team_a = CombinedSearch(*[search, search, search, search])
+    team_b = CombinedSearch(*[search, search, search, search])
+    game = TeamMatchMaker().assign_game_quality((team_a, team_b))
+
+    assert game.quality == 1.0
+
+
+def test_low_game_quality_for_high_rating_disparity(player_factory):
+    s = make_searches([100, 100, 4000, 4000, 4000, 4000, 100, 100], player_factory)
+    team_a = CombinedSearch(*[s[0], s[1], s[2], s[3]])
+    team_b = CombinedSearch(*[s[4], s[5], s[6], s[7]])
+    game = TeamMatchMaker().assign_game_quality((team_a, team_b))
+
+    assert game.quality == 0.0
+
+
+def test_low_game_quality_for_unfair_teams(player_factory):
+    s = make_searches([100, 100, 100, 100, 4000, 4000, 4000, 4000], player_factory)
+    team_a = CombinedSearch(*[s[0], s[1], s[2], s[3]])
+    team_b = CombinedSearch(*[s[4], s[5], s[6], s[7]])
+    game = TeamMatchMaker().assign_game_quality((team_a, team_b))
+
+    assert game.quality == 0.0
+
+
+@given(st_searches_list(max_players=1, min_size=6, max_size=6))
+def test_game_quality_time_bonus(s):
+    matchmaker = TeamMatchMaker()
+    team_a = CombinedSearch(*[s.pop(), s.pop(), s.pop()])
+    c1 = CombinedSearch(*[s.pop(), s.pop()])
+    team_b = CombinedSearch(*[c1, s.pop()])
+    quality_before = matchmaker.assign_game_quality((team_a, team_b)).quality
+
+    team_a.register_failed_matching_attempt()
+    team_b.register_failed_matching_attempt()
+    quality_after = matchmaker.assign_game_quality((team_a, team_b)).quality
+
+    num_newbies = sum(search.has_newbie() for search in team_a.get_original_searches())
+    num_newbies += sum(search.has_newbie() for search in team_b.get_original_searches())
+
+    assert abs(quality_before + 6 * config.TIME_BONUS + num_newbies * config.NEWBIE_BONUS - quality_after) < 0.0000001
```

Comment:
What exactly is this testing? Isn't this just duplicating the implementation in the test? Kindof like:

```python3
def add2(a, b):
    return a + b

def test_add2(a, b):
    assert add2(a, b) == a + b

Reply:
You are not wrong. I added this test to ensure that nobody accidentally breaks the time bonuses.
The assertion is a bit ugly because I get rounding errors.